From 50c127618c98a0eb7c03d43dbc0cba885640f548 Mon Sep 17 00:00:00 2001
From: Qun Zhang <qunz@codeaurora.org>
Date: Tue, 31 Jul 2018 17:32:39 +0800
Subject: [PATCH 6/9] kernel change for wlan driver on BSP 4.1.15

This change is to let wlan driver compile pass and work up on BSP 4.1.15.
propagation from msm-4.4.

Change-Id: Icc6f98cb8943f196c089428188474573fecd6453
Signed-off-by: Qun Zhang <qunz@codeaurora.org>
---
 drivers/Kconfig              |   2 +
 include/net/cfg80211.h       | 125 +++++++---
 include/uapi/linux/nl80211.h |   8 +-
 net/mac80211/util.c          |  44 ++--
 net/wireless/chan.c          |  21 +-
 net/wireless/core.c          |  14 +-
 net/wireless/core.h          |   2 +-
 net/wireless/nl80211.c       | 553 +++++++++++++++++++++++++------------------
 net/wireless/reg.c           |   6 +-
 net/wireless/sme.c           |  27 ++-
 net/wireless/util.c          | 342 ++++++++++++++++----------
 11 files changed, 709 insertions(+), 435 deletions(-)

diff --git a/drivers/Kconfig b/drivers/Kconfig
index 2b4f5b4..9565158 100644
--- a/drivers/Kconfig
+++ b/drivers/Kconfig
@@ -92,6 +92,8 @@ source "drivers/hid/Kconfig"
 
 source "drivers/usb/Kconfig"
 
+source "drivers/pci/Kconfig"
+
 source "drivers/uwb/Kconfig"
 
 source "drivers/mmc/Kconfig"
diff --git a/include/net/cfg80211.h b/include/net/cfg80211.h
index f8d6813..c114a47 100644
--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -689,6 +689,18 @@ struct cfg80211_acl_data {
 	struct mac_address mac_addrs[];
 };
 
+/*
+ * cfg80211_bitrate_mask - masks for bitrate control
+ */
+struct cfg80211_bitrate_mask {
+	struct {
+		u32 legacy;
+		u8 ht_mcs[IEEE80211_HT_MCS_MASK_LEN];
+		u16 vht_mcs[NL80211_VHT_NSS_MAX];
+		enum nl80211_txrate_gi gi;
+	} control[IEEE80211_NUM_BANDS];
+};
+
 /**
  * struct cfg80211_ap_settings - AP configuration
  *
@@ -711,6 +723,7 @@ struct cfg80211_acl_data {
  * @p2p_opp_ps: P2P opportunistic PS
  * @acl: ACL configuration used by the drivers which has support for
  *	MAC address based access control
+ * @beacon_rate: bitrate to be used for beacons
  */
 struct cfg80211_ap_settings {
 	struct cfg80211_chan_def chandef;
@@ -729,6 +742,7 @@ struct cfg80211_ap_settings {
 	u8 p2p_ctwindow;
 	bool p2p_opp_ps;
 	const struct cfg80211_acl_data *acl;
+	struct cfg80211_bitrate_mask beacon_rate;
 };
 
 /**
@@ -761,6 +775,30 @@ struct cfg80211_csa_settings {
 };
 
 /**
+ * struct iface_combination_params - input parameters for interface combinations
+ *
+ * Used to pass interface combination parameters
+ *
+ * @num_different_channels: the number of different channels we want
+ *      to use for verification
+ * @radar_detect: a bitmap where each bit corresponds to a channel
+ *      width where radar detection is needed, as in the definition of
+ *      &struct ieee80211_iface_combination.@radar_detect_widths
+ * @iftype_num: array with the number of interfaces of each interface
+ *      type.  The index is the interface type as specified in &enum
+ *      nl80211_iftype.
+ * @new_beacon_int: set this to the beacon interval of a new interface
+ *      that's not operating yet, if such is to be checked as part of
+ *      the verification
+ */
+struct iface_combination_params {
+	int num_different_channels;
+	u8 radar_detect;
+	int iftype_num[NUM_NL80211_IFTYPES];
+	u32 new_beacon_int;
+};
+
+/**
  * enum station_parameters_apply_mask - station parameter values to apply
  * @STATION_PARAM_APPLY_UAPSD: apply new uAPSD parameters (uapsd_queues, max_sp)
  * @STATION_PARAM_APPLY_CAPABILITY: apply new capability
@@ -1889,17 +1927,6 @@ enum wiphy_params_flags {
 	WIPHY_PARAM_DYN_ACK		= 1 << 5,
 };
 
-/*
- * cfg80211_bitrate_mask - masks for bitrate control
- */
-struct cfg80211_bitrate_mask {
-	struct {
-		u32 legacy;
-		u8 ht_mcs[IEEE80211_HT_MCS_MASK_LEN];
-		u16 vht_mcs[NL80211_VHT_NSS_MAX];
-		enum nl80211_txrate_gi gi;
-	} control[IEEE80211_NUM_BANDS];
-};
 /**
  * struct cfg80211_pmksa - PMK Security Association
  *
@@ -2758,6 +2785,7 @@ struct cfg80211_ops {
  * @WIPHY_FLAG_SUPPORTS_5_10_MHZ: Device supports 5 MHz and 10 MHz channels.
  * @WIPHY_FLAG_HAS_CHANNEL_SWITCH: Device supports channel switch in
  *	beaconing mode (AP, IBSS, Mesh, ...).
+ * *@WIPHY_FLAG_DFS_OFFLOAD: The driver handles all the DFS related operations.
  */
 enum wiphy_flags {
 	/* use hole at 0 */
@@ -2783,6 +2811,7 @@ enum wiphy_flags {
 	WIPHY_FLAG_HAS_REMAIN_ON_CHANNEL	= BIT(21),
 	WIPHY_FLAG_SUPPORTS_5_10_MHZ		= BIT(22),
 	WIPHY_FLAG_HAS_CHANNEL_SWITCH		= BIT(23),
+        WIPHY_FLAG_DFS_OFFLOAD			= BIT(24),
 };
 
 /**
@@ -2807,6 +2836,12 @@ struct ieee80211_iface_limit {
  *	only in special cases.
  * @radar_detect_widths: bitmap of channel widths supported for radar detection
  * @radar_detect_regions: bitmap of regions supported for radar detection
+ * @beacon_int_min_gcd: This interface combination supports different
+ *      beacon intervals.
+ *      = 0 - all beacon intervals for different interface must be same.
+ *      > 0 - any beacon interval for the interface part of this combination AND
+ *            *GCD* of all beacon intervals from beaconing interfaces of this
+ *            combination must be greater or equal to this value.
  *
  * With this structure the driver can describe which interface
  * combinations it supports concurrently.
@@ -2865,6 +2900,7 @@ struct ieee80211_iface_combination {
 	bool beacon_int_infra_match;
 	u8 radar_detect_widths;
 	u8 radar_detect_regions;
+	u32 beacon_int_min_gcd;
 };
 
 struct ieee80211_txrx_stypes {
@@ -3084,6 +3120,9 @@ struct wiphy_vendor_command {
  * @vht_capa_mod_mask:  Specify what VHT capabilities can be over-ridden.
  *	If null, then none can be over-ridden.
  *
+ * @wdev_list: the list of associated (virtual) interfaces; this list must
+ *      not be modified by the driver, but can be read with RTNL/RCU protection.
+ *
  * @max_acl_mac_addrs: Maximum number of MAC addresses that the device
  *	supports for ACL.
  *
@@ -3217,6 +3256,8 @@ struct wiphy {
 	const struct ieee80211_ht_cap *ht_capa_mod_mask;
 	const struct ieee80211_vht_cap *vht_capa_mod_mask;
 
+	struct list_head wdev_list;
+
 	/* the network namespace this phy lives in currently */
 	possible_net_t _net;
 
@@ -3827,6 +3868,32 @@ const u8 *cfg80211_find_vendor_ie(unsigned int oui, u8 oui_type,
 int regulatory_hint(struct wiphy *wiphy, const char *alpha2);
 
 /**
+ * regulatory_hint_user - hint to the wireless core a regulatory domain
+ * which the driver has received from an application
+ * @alpha2: the ISO/IEC 3166 alpha2 the driver claims its regulatory domain
+ *      should be in. If @rd is set this should be NULL. Note that if you
+ *      set this to NULL you should still set rd->alpha2 to some accepted
+ *      alpha2.
+ * @user_reg_hint_type: the type of user regulatory hint.
+ *
+ * Wireless drivers can use this function to hint to the wireless core
+ * the current regulatory domain as specified by trusted applications,
+ * it is the driver's responsibilty to estbalish which applications it
+ * trusts.
+ *
+ * The wiphy should be registered to cfg80211 prior to this call.
+ * For cfg80211 drivers this means you must first use wiphy_register(),
+ * for mac80211 drivers you must first use ieee80211_register_hw().
+ *
+ * Drivers should check the return value, its possible you can get
+ * an -ENOMEM or an -EINVAL.
+ *
+ * Return: 0 on success. -ENOMEM, -EINVAL.
+ */
+int regulatory_hint_user(const char *alpha2,
+			enum nl80211_user_reg_hint_type user_reg_hint_type);
+
+/**
  * regulatory_set_wiphy_regd - set regdom info for self managed drivers
  * @wiphy: the wireless device we want to process the regulatory domain on
  * @rd: the regulatory domain informatoin to use for this wiphy
@@ -5086,6 +5153,16 @@ void cfg80211_report_wowlan_wakeup(struct wireless_dev *wdev,
 void cfg80211_crit_proto_stopped(struct wireless_dev *wdev, gfp_t gfp);
 
 /**
+ * cfg80211_is_gratuitous_arp_unsolicited_na - packet is grat. ARP/unsol. NA
+ * @skb: the input packet, must be an ethernet frame already
+ *
+ * Return: %true if the packet is a gratuitous ARP or unsolicited NA packet.
+ * This is used to drop packets that shouldn't occur because the AP implements
+ * a proxy service.
+ */
+bool cfg80211_is_gratuitous_arp_unsolicited_na(struct sk_buff *skb);
+
+/**
  * ieee80211_get_num_supported_channels - get number of channels device has
  * @wiphy: the wiphy
  *
@@ -5097,36 +5174,20 @@ unsigned int ieee80211_get_num_supported_channels(struct wiphy *wiphy);
  * cfg80211_check_combinations - check interface combinations
  *
  * @wiphy: the wiphy
- * @num_different_channels: the number of different channels we want
- *	to use for verification
- * @radar_detect: a bitmap where each bit corresponds to a channel
- *	width where radar detection is needed, as in the definition of
- *	&struct ieee80211_iface_combination.@radar_detect_widths
- * @iftype_num: array with the numbers of interfaces of each interface
- *	type.  The index is the interface type as specified in &enum
- *	nl80211_iftype.
+ * @params: the interface combinations parameter
  *
  * This function can be called by the driver to check whether a
  * combination of interfaces and their types are allowed according to
  * the interface combinations.
  */
 int cfg80211_check_combinations(struct wiphy *wiphy,
-				const int num_different_channels,
-				const u8 radar_detect,
-				const int iftype_num[NUM_NL80211_IFTYPES]);
+				struct iface_combination_params *params);
 
 /**
  * cfg80211_iter_combinations - iterate over matching combinations
  *
  * @wiphy: the wiphy
- * @num_different_channels: the number of different channels we want
- *	to use for verification
- * @radar_detect: a bitmap where each bit corresponds to a channel
- *	width where radar detection is needed, as in the definition of
- *	&struct ieee80211_iface_combination.@radar_detect_widths
- * @iftype_num: array with the numbers of interfaces of each interface
- *	type.  The index is the interface type as specified in &enum
- *	nl80211_iftype.
+ * @params: the interface combinations parameter
  * @iter: function to call for each matching combination
  * @data: pointer to pass to iter function
  *
@@ -5135,9 +5196,7 @@ int cfg80211_check_combinations(struct wiphy *wiphy,
  * purposes.
  */
 int cfg80211_iter_combinations(struct wiphy *wiphy,
-			       const int num_different_channels,
-			       const u8 radar_detect,
-			       const int iftype_num[NUM_NL80211_IFTYPES],
+			       struct iface_combination_params *params,
 			       void (*iter)(const struct ieee80211_iface_combination *c,
 					    void *data),
 			       void *data);
diff --git a/include/uapi/linux/nl80211.h b/include/uapi/linux/nl80211.h
index 241220c..3879ba3 100644
--- a/include/uapi/linux/nl80211.h
+++ b/include/uapi/linux/nl80211.h
@@ -4027,6 +4027,9 @@ enum nl80211_iface_limit_attrs {
  *	of supported channel widths for radar detection.
  * @NL80211_IFACE_COMB_RADAR_DETECT_REGIONS: u32 attribute containing the bitmap
  *	of supported regulatory regions for radar detection.
+ * @NL80211_IFACE_COMB_BI_MIN_GCD: u32 attribute specifying the minimum GCD of
+ *      different beacon intervals supported by all the interface combinations
+ *      in this group (if not present, all beacon intervals be identical).
  * @NUM_NL80211_IFACE_COMB: number of attributes
  * @MAX_NL80211_IFACE_COMB: highest attribute number
  *
@@ -4034,8 +4037,8 @@ enum nl80211_iface_limit_attrs {
  *	limits = [ #{STA} <= 1, #{AP} <= 1 ], matching BI, channels = 1, max = 2
  *	=> allows an AP and a STA that must match BIs
  *
- *	numbers = [ #{AP, P2P-GO} <= 8 ], channels = 1, max = 8
- *	=> allows 8 of AP/GO
+ *	numbers = [ #{AP, P2P-GO} <= 8 ], BI min gcd, channels = 1, max = 8,
+ *      => allows 8 of AP/GO that can have BI gcd >= min gcd
  *
  *	numbers = [ #{STA} <= 2 ], channels = 2, max = 2
  *	=> allows two STAs on different channels
@@ -4061,6 +4064,7 @@ enum nl80211_if_combination_attrs {
 	NL80211_IFACE_COMB_NUM_CHANNELS,
 	NL80211_IFACE_COMB_RADAR_DETECT_WIDTHS,
 	NL80211_IFACE_COMB_RADAR_DETECT_REGIONS,
+	NL80211_IFACE_COMB_BI_MIN_GCD,
 
 	/* keep last */
 	NUM_NL80211_IFACE_COMB,
diff --git a/net/mac80211/util.c b/net/mac80211/util.c
index 67fec9b..5fc79e1 100644
--- a/net/mac80211/util.c
+++ b/net/mac80211/util.c
@@ -3200,10 +3200,11 @@ int ieee80211_check_combinations(struct ieee80211_sub_if_data *sdata,
 	struct ieee80211_local *local = sdata->local;
 	struct ieee80211_sub_if_data *sdata_iter;
 	enum nl80211_iftype iftype = sdata->wdev.iftype;
-	int num[NUM_NL80211_IFTYPES];
 	struct ieee80211_chanctx *ctx;
-	int num_different_channels = 0;
 	int total = 1;
+	struct iface_combination_params params = {
+		.radar_detect = radar_detect,
+        };
 
 	lockdep_assert_held(&local->chanctx_mtx);
 
@@ -3214,9 +3215,6 @@ int ieee80211_check_combinations(struct ieee80211_sub_if_data *sdata,
 		    !chandef->chan))
 		return -EINVAL;
 
-	if (chandef)
-		num_different_channels = 1;
-
 	if (WARN_ON(iftype >= NUM_NL80211_IFTYPES))
 		return -EINVAL;
 
@@ -3227,24 +3225,26 @@ int ieee80211_check_combinations(struct ieee80211_sub_if_data *sdata,
 		return 0;
 	}
 
-	memset(num, 0, sizeof(num));
+	if (chandef)
+		params.num_different_channels = 1;
 
 	if (iftype != NL80211_IFTYPE_UNSPECIFIED)
-		num[iftype] = 1;
+		params.iftype_num[iftype] = 1;
 
 	list_for_each_entry(ctx, &local->chanctx_list, list) {
 		if (ctx->replace_state == IEEE80211_CHANCTX_WILL_BE_REPLACED)
 			continue;
-		radar_detect |= ieee80211_chanctx_radar_detect(local, ctx);
+		params.radar_detect |=
+			ieee80211_chanctx_radar_detect(local, ctx);
 		if (ctx->mode == IEEE80211_CHANCTX_EXCLUSIVE) {
-			num_different_channels++;
+			params.num_different_channels++;
 			continue;
 		}
 		if (chandef && chanmode == IEEE80211_CHANCTX_SHARED &&
 		    cfg80211_chandef_compatible(chandef,
 						&ctx->conf.def))
 			continue;
-		num_different_channels++;
+		params.num_different_channels++;
 	}
 
 	list_for_each_entry_rcu(sdata_iter, &local->interfaces, list) {
@@ -3257,16 +3257,14 @@ int ieee80211_check_combinations(struct ieee80211_sub_if_data *sdata,
 		    local->hw.wiphy->software_iftypes & BIT(wdev_iter->iftype))
 			continue;
 
-		num[wdev_iter->iftype]++;
+		params.iftype_num[wdev_iter->iftype]++;
 		total++;
 	}
 
-	if (total == 1 && !radar_detect)
+	if (total == 1 && !params.radar_detect)
 		return 0;
 
-	return cfg80211_check_combinations(local->hw.wiphy,
-					   num_different_channels,
-					   radar_detect, num);
+	return cfg80211_check_combinations(local->hw.wiphy, &params);
 }
 
 static void
@@ -3282,12 +3280,10 @@ ieee80211_iter_max_chans(const struct ieee80211_iface_combination *c,
 int ieee80211_max_num_channels(struct ieee80211_local *local)
 {
 	struct ieee80211_sub_if_data *sdata;
-	int num[NUM_NL80211_IFTYPES] = {};
 	struct ieee80211_chanctx *ctx;
-	int num_different_channels = 0;
-	u8 radar_detect = 0;
 	u32 max_num_different_channels = 1;
 	int err;
+	struct iface_combination_params params = {0};
 
 	lockdep_assert_held(&local->chanctx_mtx);
 
@@ -3295,17 +3291,17 @@ int ieee80211_max_num_channels(struct ieee80211_local *local)
 		if (ctx->replace_state == IEEE80211_CHANCTX_WILL_BE_REPLACED)
 			continue;
 
-		num_different_channels++;
+		params.num_different_channels++;
 
-		radar_detect |= ieee80211_chanctx_radar_detect(local, ctx);
+		params.radar_detect |=
+                        ieee80211_chanctx_radar_detect(local, ctx);
 	}
 
 	list_for_each_entry_rcu(sdata, &local->interfaces, list)
-		num[sdata->wdev.iftype]++;
+		params.iftype_num[sdata->wdev.iftype]++;
 
-	err = cfg80211_iter_combinations(local->hw.wiphy,
-					 num_different_channels, radar_detect,
-					 num, ieee80211_iter_max_chans,
+	err = cfg80211_iter_combinations(local->hw.wiphy, &params,
+                                         ieee80211_iter_max_chans,
 					 &max_num_different_channels);
 	if (err < 0)
 		return err;
diff --git a/net/wireless/chan.c b/net/wireless/chan.c
index 7aaf741..6668846 100644
--- a/net/wireless/chan.c
+++ b/net/wireless/chan.c
@@ -319,7 +319,8 @@ static int cfg80211_get_chans_dfs_required(struct wiphy *wiphy,
 		if (!c)
 			return -EINVAL;
 
-		if (c->flags & IEEE80211_CHAN_RADAR)
+		if ((c->flags & IEEE80211_CHAN_RADAR) &&
+			!(wiphy->flags & WIPHY_FLAG_DFS_OFFLOAD))
 			return 1;
 	}
 	return 0;
@@ -479,8 +480,10 @@ static bool cfg80211_get_chans_dfs_available(struct wiphy *wiphy,
 		if (c->flags & IEEE80211_CHAN_DISABLED)
 			return false;
 
-		if ((c->flags & IEEE80211_CHAN_RADAR)  &&
-		    (c->dfs_state != NL80211_DFS_AVAILABLE))
+		/* check for radar flags */
+		if ((!(wiphy->flags & WIPHY_FLAG_DFS_OFFLOAD)) &&
+			(c->flags & IEEE80211_CHAN_RADAR) &&
+			(c->dfs_state != NL80211_DFS_AVAILABLE))
 			return false;
 	}
 
@@ -590,7 +593,15 @@ static bool cfg80211_secondary_chans_ok(struct wiphy *wiphy,
 
 	for (freq = start_freq; freq <= end_freq; freq += 20) {
 		c = ieee80211_get_channel(wiphy, freq);
-		if (!c || c->flags & prohibited_flags)
+		
+		if (!c)
+			return false;
+
+		if ((!(wiphy->flags & WIPHY_FLAG_DFS_OFFLOAD)) &&
+			(c->flags & prohibited_flags & IEEE80211_CHAN_RADAR))
+			return false;
+
+		if (c->flags & prohibited_flags & ~IEEE80211_CHAN_RADAR)
 			return false;
 	}
 
@@ -732,7 +743,7 @@ static bool cfg80211_go_permissive_chan(struct cfg80211_registered_device *rdev,
 	 * and thus fail the GO instantiation, consider only the interfaces of
 	 * the current registered device.
 	 */
-	list_for_each_entry(wdev_iter, &rdev->wdev_list, list) {
+	list_for_each_entry(wdev_iter, &rdev->wiphy.wdev_list, list) {
 		struct ieee80211_channel *other_chan = NULL;
 		int r1, r2;
 
diff --git a/net/wireless/core.c b/net/wireless/core.c
index 2a0bbd2..dcfb593 100644
--- a/net/wireless/core.c
+++ b/net/wireless/core.c
@@ -157,7 +157,7 @@ int cfg80211_switch_netns(struct cfg80211_registered_device *rdev,
 	if (!(rdev->wiphy.flags & WIPHY_FLAG_NETNS_OK))
 		return -EOPNOTSUPP;
 
-	list_for_each_entry(wdev, &rdev->wdev_list, list) {
+	list_for_each_entry(wdev, &rdev->wiphy.wdev_list, list) {
 		if (!wdev->netdev)
 			continue;
 		wdev->netdev->features &= ~NETIF_F_NETNS_LOCAL;
@@ -171,7 +171,7 @@ int cfg80211_switch_netns(struct cfg80211_registered_device *rdev,
 		/* failed -- clean up to old netns */
 		net = wiphy_net(&rdev->wiphy);
 
-		list_for_each_entry_continue_reverse(wdev, &rdev->wdev_list,
+		list_for_each_entry_continue_reverse(wdev, &rdev->wiphy.wdev_list,
 						     list) {
 			if (!wdev->netdev)
 				continue;
@@ -230,7 +230,7 @@ void cfg80211_shutdown_all_interfaces(struct wiphy *wiphy)
 
 	ASSERT_RTNL();
 
-	list_for_each_entry(wdev, &rdev->wdev_list, list) {
+	list_for_each_entry(wdev, &rdev->wiphy.wdev_list, list) {
 		if (wdev->netdev) {
 			dev_close(wdev->netdev);
 			continue;
@@ -298,7 +298,7 @@ void cfg80211_destroy_ifaces(struct cfg80211_registered_device *rdev)
 		kfree(item);
 		spin_unlock_irq(&rdev->destroy_list_lock);
 
-		list_for_each_entry_safe(wdev, tmp, &rdev->wdev_list, list) {
+		list_for_each_entry_safe(wdev, tmp, &rdev->wiphy.wdev_list, list) {
 			if (nlportid == wdev->owner_nlportid)
 				rdev_del_virtual_intf(rdev, wdev);
 		}
@@ -400,7 +400,7 @@ use_default_name:
 		dev_set_name(&rdev->wiphy.dev, PHY_NAME "%d", rdev->wiphy_idx);
 	}
 
-	INIT_LIST_HEAD(&rdev->wdev_list);
+	INIT_LIST_HEAD(&rdev->wiphy.wdev_list);
 	INIT_LIST_HEAD(&rdev->beacon_registrations);
 	spin_lock_init(&rdev->beacon_registrations_lock);
 	spin_lock_init(&rdev->bss_lock);
@@ -775,7 +775,7 @@ void wiphy_unregister(struct wiphy *wiphy)
 	nl80211_notify_wiphy(rdev, NL80211_CMD_DEL_WIPHY);
 	rdev->wiphy.registered = false;
 
-	WARN_ON(!list_empty(&rdev->wdev_list));
+	WARN_ON(!list_empty(&rdev->wiphy.wdev_list));
 
 	/*
 	 * First remove the hardware from everywhere, this makes
@@ -995,7 +995,7 @@ static int cfg80211_netdev_notifier_call(struct notifier_block *nb,
 		spin_lock_init(&wdev->mgmt_registrations_lock);
 
 		wdev->identifier = ++rdev->wdev_id;
-		list_add_rcu(&wdev->list, &rdev->wdev_list);
+		list_add_rcu(&wdev->list, &rdev->wiphy.wdev_list);
 		rdev->devlist_generation++;
 		/* can only change netns with wiphy */
 		dev->features |= NETIF_F_NETNS_LOCAL;
diff --git a/net/wireless/core.h b/net/wireless/core.h
index 801cd49..29e66e5 100644
--- a/net/wireless/core.h
+++ b/net/wireless/core.h
@@ -463,7 +463,7 @@ int ieee80211_get_ratemask(struct ieee80211_supported_band *sband,
 			   u32 *mask);
 
 int cfg80211_validate_beacon_int(struct cfg80211_registered_device *rdev,
-				 u32 beacon_int);
+				 enum nl80211_iftype iftype, u32 beacon_int);
 
 void cfg80211_update_iface_num(struct cfg80211_registered_device *rdev,
 			       enum nl80211_iftype iftype, int num);
diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index 04b6f3f..f2f6c73 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -3,6 +3,7 @@
  *
  * Copyright 2006-2010	Johannes Berg <johannes@sipsolutions.net>
  * Copyright 2013-2014  Intel Mobile Communications GmbH
+ * Copyright 2015       Intel Deutschland GmbH
  */
 
 #include <linux/if.h>
@@ -102,7 +103,7 @@ __cfg80211_wdev_from_attrs(struct net *netns, struct nlattr **attrs)
 		if (have_wdev_id && rdev->wiphy_idx != wiphy_idx)
 			continue;
 
-		list_for_each_entry(wdev, &rdev->wdev_list, list) {
+		list_for_each_entry(wdev, &rdev->wiphy.wdev_list, list) {
 			if (have_ifidx && wdev->netdev &&
 			    wdev->netdev->ifindex == ifidx) {
 				result = wdev;
@@ -148,7 +149,7 @@ __cfg80211_rdev_from_attrs(struct net *netns, struct nlattr **attrs)
 		tmp = cfg80211_rdev_by_wiphy_idx(wdev_id >> 32);
 		if (tmp) {
 			/* make sure wdev exists */
-			list_for_each_entry(wdev, &tmp->wdev_list, list) {
+			list_for_each_entry(wdev, &tmp->wiphy.wdev_list, list) {
 				if (wdev->identifier != (u32)wdev_id)
 					continue;
 				found = true;
@@ -516,7 +517,7 @@ static int nl80211_prepare_wdev_dump(struct sk_buff *skb,
 		*rdev = wiphy_to_rdev(wiphy);
 		*wdev = NULL;
 
-		list_for_each_entry(tmp, &(*rdev)->wdev_list, list) {
+		list_for_each_entry(tmp, &(*rdev)->wiphy.wdev_list, list) {
 			if (tmp->identifier == cb->args[1]) {
 				*wdev = tmp;
 				break;
@@ -995,6 +996,11 @@ static int nl80211_put_iface_combinations(struct wiphy *wiphy,
 		     nla_put_u32(msg, NL80211_IFACE_COMB_RADAR_DETECT_REGIONS,
 				c->radar_detect_regions)))
 			goto nla_put_failure;
+                        
+		if (c->beacon_int_min_gcd &&
+			nla_put_u32(msg, NL80211_IFACE_COMB_BI_MIN_GCD,
+			c->beacon_int_min_gcd))
+			goto nla_put_failure;
 
 		nla_nest_end(msg, nl_combi);
 	}
@@ -2433,7 +2439,7 @@ static int nl80211_dump_interface(struct sk_buff *skb, struct netlink_callback *
 		}
 		if_idx = 0;
 
-		list_for_each_entry(wdev, &rdev->wdev_list, list) {
+		list_for_each_entry(wdev, &rdev->wiphy.wdev_list, list) {
 			if (if_idx < if_start) {
 				if_idx++;
 				continue;
@@ -2705,7 +2711,7 @@ static int nl80211_new_interface(struct sk_buff *skb, struct genl_info *info)
 		spin_lock_init(&wdev->mgmt_registrations_lock);
 
 		wdev->identifier = ++rdev->wdev_id;
-		list_add_rcu(&wdev->list, &rdev->wdev_list);
+		list_add_rcu(&wdev->list, &rdev->wiphy.wdev_list);
 		rdev->devlist_generation++;
 		break;
 	default:
@@ -3178,6 +3184,279 @@ static int nl80211_set_mac_acl(struct sk_buff *skb, struct genl_info *info)
 	return err;
 }
 
+static u32 rateset_to_mask(struct ieee80211_supported_band *sband,
+                           u8 *rates, u8 rates_len)
+{
+        u8 i;
+        u32 mask = 0;
+
+        for (i = 0; i < rates_len; i++) {
+                int rate = (rates[i] & 0x7f) * 5;
+                int ridx;
+
+                for (ridx = 0; ridx < sband->n_bitrates; ridx++) {
+                        struct ieee80211_rate *srate =
+                                &sband->bitrates[ridx];
+                        if (rate == srate->bitrate) {
+                                mask |= 1 << ridx;
+                                break;
+                        }
+                }
+                if (ridx == sband->n_bitrates)
+                        return 0; /* rate not found */
+        }
+
+        return mask;
+}
+
+static bool ht_rateset_to_mask(struct ieee80211_supported_band *sband,
+                               u8 *rates, u8 rates_len,
+                               u8 mcs[IEEE80211_HT_MCS_MASK_LEN])
+{
+        u8 i;
+
+        memset(mcs, 0, IEEE80211_HT_MCS_MASK_LEN);
+
+        for (i = 0; i < rates_len; i++) {
+                int ridx, rbit;
+
+                ridx = rates[i] / 8;
+                rbit = BIT(rates[i] % 8);
+
+                /* check validity */
+                if ((ridx < 0) || (ridx >= IEEE80211_HT_MCS_MASK_LEN))
+                        return false;
+
+                /* check availability */
+                if (sband->ht_cap.mcs.rx_mask[ridx] & rbit)
+                        mcs[ridx] |= rbit;
+                else
+                        return false;
+        }
+
+        return true;
+}
+
+static u16 vht_mcs_map_to_mcs_mask(u8 vht_mcs_map)
+{
+        u16 mcs_mask = 0;
+
+        switch (vht_mcs_map) {
+        case IEEE80211_VHT_MCS_NOT_SUPPORTED:
+                break;
+        case IEEE80211_VHT_MCS_SUPPORT_0_7:
+                mcs_mask = 0x00FF;
+                break;
+        case IEEE80211_VHT_MCS_SUPPORT_0_8:
+                mcs_mask = 0x01FF;
+                break;
+        case IEEE80211_VHT_MCS_SUPPORT_0_9:
+                mcs_mask = 0x03FF;
+                break;
+        default:
+                break;
+        }
+
+        return mcs_mask;
+}
+
+static void vht_build_mcs_mask(u16 vht_mcs_map,
+                               u16 vht_mcs_mask[NL80211_VHT_NSS_MAX])
+{
+        u8 nss;
+
+        for (nss = 0; nss < NL80211_VHT_NSS_MAX; nss++) {
+                vht_mcs_mask[nss] = vht_mcs_map_to_mcs_mask(vht_mcs_map & 0x03);
+                vht_mcs_map >>= 2;
+        }
+}
+
+static bool vht_set_mcs_mask(struct ieee80211_supported_band *sband,
+                             struct nl80211_txrate_vht *txrate,
+                             u16 mcs[NL80211_VHT_NSS_MAX])
+{
+        u16 tx_mcs_map = le16_to_cpu(sband->vht_cap.vht_mcs.tx_mcs_map);
+        u16 tx_mcs_mask[NL80211_VHT_NSS_MAX] = {};
+        u8 i;
+
+        if (!sband->vht_cap.vht_supported)
+                return false;
+
+        memset(mcs, 0, sizeof(u16) * NL80211_VHT_NSS_MAX);
+
+        /* Build vht_mcs_mask from VHT capabilities */
+        vht_build_mcs_mask(tx_mcs_map, tx_mcs_mask);
+
+        for (i = 0; i < NL80211_VHT_NSS_MAX; i++) {
+                if ((tx_mcs_mask[i] & txrate->mcs[i]) == txrate->mcs[i])
+                        mcs[i] = txrate->mcs[i];
+                else
+                        return false;
+        }
+
+        return true;
+}
+
+static const struct nla_policy nl80211_txattr_policy[NL80211_TXRATE_MAX + 1] = {
+        [NL80211_TXRATE_LEGACY] = { .type = NLA_BINARY,
+                                    .len = NL80211_MAX_SUPP_RATES },
+        [NL80211_TXRATE_HT] = { .type = NLA_BINARY,
+                                .len = NL80211_MAX_SUPP_HT_RATES },
+        [NL80211_TXRATE_VHT] = { .len = sizeof(struct nl80211_txrate_vht)},
+        [NL80211_TXRATE_GI] = { .type = NLA_U8 },
+};
+
+static int nl80211_parse_tx_bitrate_mask(struct genl_info *info,
+                                         struct cfg80211_bitrate_mask *mask)
+{
+        struct nlattr *tb[NL80211_TXRATE_MAX + 1];
+        struct cfg80211_registered_device *rdev = info->user_ptr[0];
+        int rem, i;
+        struct nlattr *tx_rates;
+        struct ieee80211_supported_band *sband;
+        u16 vht_tx_mcs_map;
+
+        memset(mask, 0, sizeof(*mask));
+        /* Default to all rates enabled */
+        for (i = 0; i < IEEE80211_NUM_BANDS; i++) {
+                sband = rdev->wiphy.bands[i];
+
+                if (!sband)
+                        continue;
+
+                mask->control[i].legacy = (1 << sband->n_bitrates) - 1;
+                memcpy(mask->control[i].ht_mcs,
+                       sband->ht_cap.mcs.rx_mask,
+                       sizeof(mask->control[i].ht_mcs));
+
+                if (!sband->vht_cap.vht_supported)
+                        continue;
+
+                vht_tx_mcs_map = le16_to_cpu(sband->vht_cap.vht_mcs.tx_mcs_map);
+                vht_build_mcs_mask(vht_tx_mcs_map, mask->control[i].vht_mcs);
+        }
+
+        /* if no rates are given set it back to the defaults */
+        if (!info->attrs[NL80211_ATTR_TX_RATES])
+                goto out;
+
+        /* The nested attribute uses enum nl80211_band as the index. This maps
+         * directly to the enum nl80211_band values used in cfg80211.
+         */
+        BUILD_BUG_ON(NL80211_MAX_SUPP_HT_RATES > IEEE80211_HT_MCS_MASK_LEN * 8);
+        nla_for_each_nested(tx_rates, info->attrs[NL80211_ATTR_TX_RATES], rem) {
+                enum ieee80211_band band = nla_type(tx_rates);
+                int err;
+
+                if (band < 0 || band >= IEEE80211_NUM_BANDS)
+                        return -EINVAL;
+                sband = rdev->wiphy.bands[band];
+                if (sband == NULL)
+                        return -EINVAL;
+                err = nla_parse(tb, NL80211_TXRATE_MAX, nla_data(tx_rates),
+                                nla_len(tx_rates), nl80211_txattr_policy);
+                if (err)
+                        return err;
+		if (tb[NL80211_TXRATE_LEGACY]) {
+                        mask->control[band].legacy = rateset_to_mask(
+                                sband,
+                                nla_data(tb[NL80211_TXRATE_LEGACY]),
+                                nla_len(tb[NL80211_TXRATE_LEGACY]));
+                        if ((mask->control[band].legacy == 0) &&
+                            nla_len(tb[NL80211_TXRATE_LEGACY]))
+                                return -EINVAL;
+                }
+                if (tb[NL80211_TXRATE_HT]) {
+                        if (!ht_rateset_to_mask(
+                                        sband,
+                                        nla_data(tb[NL80211_TXRATE_HT]),
+                                        nla_len(tb[NL80211_TXRATE_HT]),
+                                        mask->control[band].ht_mcs))
+                                return -EINVAL;
+                }
+                if (tb[NL80211_TXRATE_VHT]) {
+                        if (!vht_set_mcs_mask(
+                                        sband,
+                                        nla_data(tb[NL80211_TXRATE_VHT]),
+                                        mask->control[band].vht_mcs))
+                                return -EINVAL;
+                }
+                if (tb[NL80211_TXRATE_GI]) {
+                        mask->control[band].gi =
+                                nla_get_u8(tb[NL80211_TXRATE_GI]);
+                        if (mask->control[band].gi > NL80211_TXRATE_FORCE_LGI)
+                                return -EINVAL;
+                }
+
+                if (mask->control[band].legacy == 0) {
+                        /* don't allow empty legacy rates if HT or VHT
+                         * are not even supported.
+                         */
+                      if (!(rdev->wiphy.bands[band]->ht_cap.ht_supported ||
+                              rdev->wiphy.bands[band]->vht_cap.vht_supported))
+                                return -EINVAL;
+
+                        for (i = 0; i < IEEE80211_HT_MCS_MASK_LEN; i++)
+                                if (mask->control[band].ht_mcs[i])
+                                        goto out;
+
+                        for (i = 0; i < NL80211_VHT_NSS_MAX; i++)
+                                if (mask->control[band].vht_mcs[i])
+                                        goto out;
+
+                        /* legacy and mcs rates may not be both empty */
+                        return -EINVAL;
+                }
+        }
+
+out:
+        return 0;
+}
+
+static int validate_beacon_tx_rate(struct cfg80211_ap_settings *params)
+{
+        u32 rate,count_ht, count_vht, i;
+        enum nl80211_band band;
+
+        band = params->chandef.chan->band;
+        rate = params->beacon_rate.control[band].legacy;
+
+        /* Allow only one rate */
+        if (hweight32(rate) > 1)
+                return -EINVAL;
+
+        count_ht = 0;
+        for (i = 0; i < IEEE80211_HT_MCS_MASK_LEN; i++) {
+                if (hweight8(params->beacon_rate.control[band].ht_mcs[i]) > 1) {
+                        return -EINVAL;
+                } else if (params->beacon_rate.control[band].ht_mcs[i]) {
+                        count_ht++;
+                        if (count_ht > 1)
+                                return -EINVAL;
+                }
+                if (count_ht && rate)
+                        return -EINVAL;
+        }
+
+        count_vht = 0;
+        for (i = 0; i < NL80211_VHT_NSS_MAX; i++) {
+                if (hweight16(params->beacon_rate.control[band].vht_mcs[i]) > 1) {
+                        return -EINVAL;
+                } else if (params->beacon_rate.control[band].vht_mcs[i]) {
+                        count_vht++;
+                        if (count_vht > 1)
+                                return -EINVAL;
+                }
+                if (count_vht && rate)
+                        return -EINVAL;
+        }
+
+        if ((count_ht && count_vht) || (!rate && !count_ht && !count_vht))
+                return -EINVAL;
+
+        return 0;
+}
+
 static int nl80211_parse_beacon(struct nlattr *attrs[],
 				struct cfg80211_beacon_data *bcn)
 {
@@ -3241,7 +3520,7 @@ static bool nl80211_get_ap_channel(struct cfg80211_registered_device *rdev,
 	struct wireless_dev *wdev;
 	bool ret = false;
 
-	list_for_each_entry(wdev, &rdev->wdev_list, list) {
+	list_for_each_entry(wdev, &rdev->wiphy.wdev_list, list) {
 		if (wdev->iftype != NL80211_IFTYPE_AP &&
 		    wdev->iftype != NL80211_IFTYPE_P2P_GO)
 			continue;
@@ -3316,7 +3595,8 @@ static int nl80211_start_ap(struct sk_buff *skb, struct genl_info *info)
 	params.dtim_period =
 		nla_get_u32(info->attrs[NL80211_ATTR_DTIM_PERIOD]);
 
-	err = cfg80211_validate_beacon_int(rdev, params.beacon_interval);
+	err = cfg80211_validate_beacon_int(rdev, dev->ieee80211_ptr->iftype,
+						params.beacon_interval);
 	if (err)
 		return err;
 
@@ -3407,6 +3687,16 @@ static int nl80211_start_ap(struct sk_buff *skb, struct genl_info *info)
 				     wdev->iftype))
 		return -EINVAL;
 
+	if (info->attrs[NL80211_ATTR_TX_RATES]) {
+                err = nl80211_parse_tx_bitrate_mask(info, &params.beacon_rate);
+                if (err)
+                        return err;
+
+                err = validate_beacon_tx_rate(&params);
+                if (err)
+                        return err;
+        }
+
 	if (info->attrs[NL80211_ATTR_SMPS_MODE]) {
 		params.smps_mode =
 			nla_get_u8(info->attrs[NL80211_ATTR_SMPS_MODE]);
@@ -6321,6 +6611,9 @@ static int nl80211_start_radar_detection(struct sk_buff *skb,
 	if (err)
 		return err;
 
+	if (rdev->wiphy.flags & WIPHY_FLAG_DFS_OFFLOAD)
+		return -EOPNOTSUPP;
+
 	if (netif_carrier_ok(dev))
 		return -EBUSY;
 
@@ -7242,12 +7535,14 @@ static int nl80211_join_ibss(struct sk_buff *skb, struct genl_info *info)
 
 	ibss.beacon_interval = 100;
 
-	if (info->attrs[NL80211_ATTR_BEACON_INTERVAL]) {
+	if (info->attrs[NL80211_ATTR_BEACON_INTERVAL])
 		ibss.beacon_interval =
 			nla_get_u32(info->attrs[NL80211_ATTR_BEACON_INTERVAL]);
-		if (ibss.beacon_interval < 1 || ibss.beacon_interval > 10000)
-			return -EINVAL;
-	}
+        
+        err = cfg80211_validate_beacon_int(rdev, NL80211_IFTYPE_ADHOC,
+                                           ibss.beacon_interval);
+        if (err)
+                return err;
 
 	if (!rdev->ops->join_ibss)
 		return -EOPNOTSUPP;
@@ -8039,237 +8334,21 @@ static int nl80211_cancel_remain_on_channel(struct sk_buff *skb,
 	return rdev_cancel_remain_on_channel(rdev, wdev, cookie);
 }
 
-static u32 rateset_to_mask(struct ieee80211_supported_band *sband,
-			   u8 *rates, u8 rates_len)
-{
-	u8 i;
-	u32 mask = 0;
-
-	for (i = 0; i < rates_len; i++) {
-		int rate = (rates[i] & 0x7f) * 5;
-		int ridx;
-		for (ridx = 0; ridx < sband->n_bitrates; ridx++) {
-			struct ieee80211_rate *srate =
-				&sband->bitrates[ridx];
-			if (rate == srate->bitrate) {
-				mask |= 1 << ridx;
-				break;
-			}
-		}
-		if (ridx == sband->n_bitrates)
-			return 0; /* rate not found */
-	}
-
-	return mask;
-}
-
-static bool ht_rateset_to_mask(struct ieee80211_supported_band *sband,
-			       u8 *rates, u8 rates_len,
-			       u8 mcs[IEEE80211_HT_MCS_MASK_LEN])
-{
-	u8 i;
-
-	memset(mcs, 0, IEEE80211_HT_MCS_MASK_LEN);
-
-	for (i = 0; i < rates_len; i++) {
-		int ridx, rbit;
-
-		ridx = rates[i] / 8;
-		rbit = BIT(rates[i] % 8);
-
-		/* check validity */
-		if ((ridx < 0) || (ridx >= IEEE80211_HT_MCS_MASK_LEN))
-			return false;
-
-		/* check availability */
-		if (sband->ht_cap.mcs.rx_mask[ridx] & rbit)
-			mcs[ridx] |= rbit;
-		else
-			return false;
-	}
-
-	return true;
-}
-
-static u16 vht_mcs_map_to_mcs_mask(u8 vht_mcs_map)
-{
-	u16 mcs_mask = 0;
-
-	switch (vht_mcs_map) {
-	case IEEE80211_VHT_MCS_NOT_SUPPORTED:
-		break;
-	case IEEE80211_VHT_MCS_SUPPORT_0_7:
-		mcs_mask = 0x00FF;
-		break;
-	case IEEE80211_VHT_MCS_SUPPORT_0_8:
-		mcs_mask = 0x01FF;
-		break;
-	case IEEE80211_VHT_MCS_SUPPORT_0_9:
-		mcs_mask = 0x03FF;
-		break;
-	default:
-		break;
-	}
-
-	return mcs_mask;
-}
-
-static void vht_build_mcs_mask(u16 vht_mcs_map,
-			       u16 vht_mcs_mask[NL80211_VHT_NSS_MAX])
-{
-	u8 nss;
-
-	for (nss = 0; nss < NL80211_VHT_NSS_MAX; nss++) {
-		vht_mcs_mask[nss] = vht_mcs_map_to_mcs_mask(vht_mcs_map & 0x03);
-		vht_mcs_map >>= 2;
-	}
-}
-
-static bool vht_set_mcs_mask(struct ieee80211_supported_band *sband,
-			     struct nl80211_txrate_vht *txrate,
-			     u16 mcs[NL80211_VHT_NSS_MAX])
-{
-	u16 tx_mcs_map = le16_to_cpu(sband->vht_cap.vht_mcs.tx_mcs_map);
-	u16 tx_mcs_mask[NL80211_VHT_NSS_MAX] = {};
-	u8 i;
-
-	if (!sband->vht_cap.vht_supported)
-		return false;
-
-	memset(mcs, 0, sizeof(u16) * NL80211_VHT_NSS_MAX);
-
-	/* Build vht_mcs_mask from VHT capabilities */
-	vht_build_mcs_mask(tx_mcs_map, tx_mcs_mask);
-
-	for (i = 0; i < NL80211_VHT_NSS_MAX; i++) {
-		if ((tx_mcs_mask[i] & txrate->mcs[i]) == txrate->mcs[i])
-			mcs[i] = txrate->mcs[i];
-		else
-			return false;
-	}
-
-	return true;
-}
-
-static const struct nla_policy nl80211_txattr_policy[NL80211_TXRATE_MAX + 1] = {
-	[NL80211_TXRATE_LEGACY] = { .type = NLA_BINARY,
-				    .len = NL80211_MAX_SUPP_RATES },
-	[NL80211_TXRATE_HT] = { .type = NLA_BINARY,
-				.len = NL80211_MAX_SUPP_HT_RATES },
-	[NL80211_TXRATE_VHT] = { .len = sizeof(struct nl80211_txrate_vht)},
-	[NL80211_TXRATE_GI] = { .type = NLA_U8 },
-};
-
 static int nl80211_set_tx_bitrate_mask(struct sk_buff *skb,
 				       struct genl_info *info)
 {
-	struct nlattr *tb[NL80211_TXRATE_MAX + 1];
-	struct cfg80211_registered_device *rdev = info->user_ptr[0];
 	struct cfg80211_bitrate_mask mask;
-	int rem, i;
+	struct cfg80211_registered_device *rdev = info->user_ptr[0];
 	struct net_device *dev = info->user_ptr[1];
-	struct nlattr *tx_rates;
-	struct ieee80211_supported_band *sband;
-	u16 vht_tx_mcs_map;
+	int err;
 
 	if (!rdev->ops->set_bitrate_mask)
 		return -EOPNOTSUPP;
 
-	memset(&mask, 0, sizeof(mask));
-	/* Default to all rates enabled */
-	for (i = 0; i < IEEE80211_NUM_BANDS; i++) {
-		sband = rdev->wiphy.bands[i];
-
-		if (!sband)
-			continue;
-
-		mask.control[i].legacy = (1 << sband->n_bitrates) - 1;
-		memcpy(mask.control[i].ht_mcs,
-		       sband->ht_cap.mcs.rx_mask,
-		       sizeof(mask.control[i].ht_mcs));
-
-		if (!sband->vht_cap.vht_supported)
-			continue;
-
-		vht_tx_mcs_map = le16_to_cpu(sband->vht_cap.vht_mcs.tx_mcs_map);
-		vht_build_mcs_mask(vht_tx_mcs_map, mask.control[i].vht_mcs);
-	}
-
-	/* if no rates are given set it back to the defaults */
-	if (!info->attrs[NL80211_ATTR_TX_RATES])
-		goto out;
-
-	/*
-	 * The nested attribute uses enum nl80211_band as the index. This maps
-	 * directly to the enum ieee80211_band values used in cfg80211.
-	 */
-	BUILD_BUG_ON(NL80211_MAX_SUPP_HT_RATES > IEEE80211_HT_MCS_MASK_LEN * 8);
-	nla_for_each_nested(tx_rates, info->attrs[NL80211_ATTR_TX_RATES], rem) {
-		enum ieee80211_band band = nla_type(tx_rates);
-		int err;
-
-		if (band < 0 || band >= IEEE80211_NUM_BANDS)
-			return -EINVAL;
-		sband = rdev->wiphy.bands[band];
-		if (sband == NULL)
-			return -EINVAL;
-		err = nla_parse(tb, NL80211_TXRATE_MAX, nla_data(tx_rates),
-				nla_len(tx_rates), nl80211_txattr_policy);
-		if (err)
-			return err;
-		if (tb[NL80211_TXRATE_LEGACY]) {
-			mask.control[band].legacy = rateset_to_mask(
-				sband,
-				nla_data(tb[NL80211_TXRATE_LEGACY]),
-				nla_len(tb[NL80211_TXRATE_LEGACY]));
-			if ((mask.control[band].legacy == 0) &&
-			    nla_len(tb[NL80211_TXRATE_LEGACY]))
-				return -EINVAL;
-		}
-		if (tb[NL80211_TXRATE_HT]) {
-			if (!ht_rateset_to_mask(
-					sband,
-					nla_data(tb[NL80211_TXRATE_HT]),
-					nla_len(tb[NL80211_TXRATE_HT]),
-					mask.control[band].ht_mcs))
-				return -EINVAL;
-		}
-		if (tb[NL80211_TXRATE_VHT]) {
-			if (!vht_set_mcs_mask(
-					sband,
-					nla_data(tb[NL80211_TXRATE_VHT]),
-					mask.control[band].vht_mcs))
-				return -EINVAL;
-		}
-		if (tb[NL80211_TXRATE_GI]) {
-			mask.control[band].gi =
-				nla_get_u8(tb[NL80211_TXRATE_GI]);
-			if (mask.control[band].gi > NL80211_TXRATE_FORCE_LGI)
-				return -EINVAL;
-		}
-
-		if (mask.control[band].legacy == 0) {
-			/* don't allow empty legacy rates if HT or VHT
-			 * are not even supported.
-			 */
-			if (!(rdev->wiphy.bands[band]->ht_cap.ht_supported ||
-			      rdev->wiphy.bands[band]->vht_cap.vht_supported))
-				return -EINVAL;
-
-			for (i = 0; i < IEEE80211_HT_MCS_MASK_LEN; i++)
-				if (mask.control[band].ht_mcs[i])
-					goto out;
-
-			for (i = 0; i < NL80211_VHT_NSS_MAX; i++)
-				if (mask.control[band].vht_mcs[i])
-					goto out;
-
-			/* legacy and mcs rates may not be both empty */
-			return -EINVAL;
-		}
-	}
+	err = nl80211_parse_tx_bitrate_mask(info, &mask);
+	if (err)
+		return err;
 
-out:
 	return rdev_set_bitrate_mask(rdev, dev, NULL, &mask);
 }
 
@@ -8688,9 +8767,11 @@ static int nl80211_join_mesh(struct sk_buff *skb, struct genl_info *info)
 	if (info->attrs[NL80211_ATTR_BEACON_INTERVAL]) {
 		setup.beacon_interval =
 			nla_get_u32(info->attrs[NL80211_ATTR_BEACON_INTERVAL]);
-		if (setup.beacon_interval < 10 ||
-		    setup.beacon_interval > 10000)
-			return -EINVAL;
+		err = cfg80211_validate_beacon_int(rdev,
+                                                   NL80211_IFTYPE_MESH_POINT,
+                                                   setup.beacon_interval);
+                if (err)
+                        return err;
 	}
 
 	if (info->attrs[NL80211_ATTR_DTIM_PERIOD]) {
@@ -12792,7 +12873,7 @@ static int nl80211_netlink_notify(struct notifier_block * nb,
 		    sched_scan_req->owner_nlportid == notify->portid)
 			schedule_scan_stop = true;
 
-		list_for_each_entry_rcu(wdev, &rdev->wdev_list, list) {
+		list_for_each_entry_rcu(wdev, &rdev->wiphy.wdev_list, list) {
 			cfg80211_mlme_unregister_socket(wdev, notify->portid);
 
 			if (wdev->owner_nlportid == notify->portid)
diff --git a/net/wireless/reg.c b/net/wireless/reg.c
index 88d6178..3bca187 100644
--- a/net/wireless/reg.c
+++ b/net/wireless/reg.c
@@ -1612,7 +1612,7 @@ static void reg_leave_invalid_chans(struct wiphy *wiphy)
 
 	ASSERT_RTNL();
 
-	list_for_each_entry(wdev, &rdev->wdev_list, list)
+	list_for_each_entry(wdev, &rdev->wiphy.wdev_list, list)
 		if (!reg_wdev_chan_valid(wiphy, wdev))
 			cfg80211_leave(rdev, wdev);
 }
@@ -2311,6 +2311,8 @@ int regulatory_hint_user(const char *alpha2,
 	return 0;
 }
 
+EXPORT_SYMBOL(regulatory_hint_user);
+
 int regulatory_hint_indoor(bool is_indoor, u32 portid)
 {
 	spin_lock(&reg_indoor_lock);
@@ -2584,7 +2586,7 @@ static void restore_regulatory_settings(bool reset_user)
 	 * settings, user regulatory settings takes precedence.
 	 */
 	if (is_an_alpha2(alpha2))
-		regulatory_hint_user(user_alpha2, NL80211_USER_REG_HINT_USER);
+		regulatory_hint_user(alpha2, NL80211_USER_REG_HINT_USER);
 
 	spin_lock(&reg_requests_lock);
 	list_splice_tail_init(&tmp_reg_req_list, &reg_requests_list);
diff --git a/net/wireless/sme.c b/net/wireless/sme.c
index d11454f..df73535 100644
--- a/net/wireless/sme.c
+++ b/net/wireless/sme.c
@@ -47,6 +47,29 @@ struct cfg80211_conn {
 	bool auto_auth, prev_bssid_valid;
 };
 
+static bool cfg80211_is_all_countryie_ignore(void)
+{
+        struct cfg80211_registered_device *rdev;
+        struct wireless_dev *wdev;
+        bool is_all_countryie_ignore = true;
+
+        list_for_each_entry(rdev, &cfg80211_rdev_list, list) {
+                list_for_each_entry(wdev, &rdev->wiphy.wdev_list, list) {
+                        wdev_lock(wdev);
+                        if (!(wdev->wiphy->regulatory_flags &
+                                REGULATORY_COUNTRY_IE_IGNORE)) {
+                                is_all_countryie_ignore = false;
+                                wdev_unlock(wdev);
+                                goto out;
+                        }
+                        wdev_unlock(wdev);
+                }
+        }
+
+out:
+        return is_all_countryie_ignore;
+}
+
 static void cfg80211_sme_free(struct wireless_dev *wdev)
 {
 	if (!wdev->conn)
@@ -605,7 +628,7 @@ static bool cfg80211_is_all_idle(void)
 	 * count as new regulatory hints.
 	 */
 	list_for_each_entry(rdev, &cfg80211_rdev_list, list) {
-		list_for_each_entry(wdev, &rdev->wdev_list, list) {
+		list_for_each_entry(wdev, &rdev->wiphy.wdev_list, list) {
 			wdev_lock(wdev);
 			if (wdev->conn || wdev->current_bss)
 				is_all_idle = false;
@@ -619,7 +642,7 @@ static bool cfg80211_is_all_idle(void)
 static void disconnect_work(struct work_struct *work)
 {
 	rtnl_lock();
-	if (cfg80211_is_all_idle())
+	if (cfg80211_is_all_idle() && !cfg80211_is_all_countryie_ignore())
 		regulatory_hint_disconnect();
 	rtnl_unlock();
 }
diff --git a/net/wireless/util.c b/net/wireless/util.c
index 7e4e3ff..5d4a0cf 100644
--- a/net/wireless/util.c
+++ b/net/wireless/util.c
@@ -13,6 +13,7 @@
 #include <net/dsfield.h>
 #include <linux/if_vlan.h>
 #include <linux/mpls.h>
+#include <linux/gcd.h>
 #include "core.h"
 #include "rdev-ops.h"
 
@@ -912,7 +913,7 @@ void cfg80211_process_rdev_events(struct cfg80211_registered_device *rdev)
 
 	ASSERT_RTNL();
 
-	list_for_each_entry(wdev, &rdev->wdev_list, list)
+	list_for_each_entry(wdev, &rdev->wiphy.wdev_list, list)
 		cfg80211_process_wdev_events(wdev);
 }
 
@@ -1484,111 +1485,160 @@ bool ieee80211_chandef_to_operating_class(struct cfg80211_chan_def *chandef,
 }
 EXPORT_SYMBOL(ieee80211_chandef_to_operating_class);
 
-int cfg80211_validate_beacon_int(struct cfg80211_registered_device *rdev,
-				 u32 beacon_int)
+static void cfg80211_calculate_bi_data(struct wiphy *wiphy, u32 new_beacon_int,
+                                       u32 *beacon_int_gcd,
+                                       bool *beacon_int_different)
 {
-	struct wireless_dev *wdev;
-	int res = 0;
-
-	if (!beacon_int)
-		return -EINVAL;
+        struct wireless_dev *wdev;
 
-	list_for_each_entry(wdev, &rdev->wdev_list, list) {
-		if (!wdev->beacon_interval)
-			continue;
-		if (wdev->beacon_interval != beacon_int) {
-			res = -EINVAL;
-			break;
-		}
-	}
+        *beacon_int_gcd = 0;
+        *beacon_int_different = false;
 
-	return res;
-}
+        list_for_each_entry(wdev, &wiphy->wdev_list, list) {
+                if (!wdev->beacon_interval)
+                        continue;
 
-int cfg80211_iter_combinations(struct wiphy *wiphy,
-			       const int num_different_channels,
-			       const u8 radar_detect,
-			       const int iftype_num[NUM_NL80211_IFTYPES],
-			       void (*iter)(const struct ieee80211_iface_combination *c,
-					    void *data),
-			       void *data)
-{
-	const struct ieee80211_regdomain *regdom;
-	enum nl80211_dfs_regions region = 0;
-	int i, j, iftype;
-	int num_interfaces = 0;
-	u32 used_iftypes = 0;
-
-	if (radar_detect) {
-		rcu_read_lock();
-		regdom = rcu_dereference(cfg80211_regdomain);
-		if (regdom)
-			region = regdom->dfs_region;
-		rcu_read_unlock();
-	}
-
-	for (iftype = 0; iftype < NUM_NL80211_IFTYPES; iftype++) {
-		num_interfaces += iftype_num[iftype];
-		if (iftype_num[iftype] > 0 &&
-		    !(wiphy->software_iftypes & BIT(iftype)))
-			used_iftypes |= BIT(iftype);
-	}
-
-	for (i = 0; i < wiphy->n_iface_combinations; i++) {
-		const struct ieee80211_iface_combination *c;
-		struct ieee80211_iface_limit *limits;
-		u32 all_iftypes = 0;
-
-		c = &wiphy->iface_combinations[i];
-
-		if (num_interfaces > c->max_interfaces)
-			continue;
-		if (num_different_channels > c->num_different_channels)
-			continue;
-
-		limits = kmemdup(c->limits, sizeof(limits[0]) * c->n_limits,
-				 GFP_KERNEL);
-		if (!limits)
-			return -ENOMEM;
-
-		for (iftype = 0; iftype < NUM_NL80211_IFTYPES; iftype++) {
-			if (wiphy->software_iftypes & BIT(iftype))
-				continue;
-			for (j = 0; j < c->n_limits; j++) {
-				all_iftypes |= limits[j].types;
-				if (!(limits[j].types & BIT(iftype)))
-					continue;
-				if (limits[j].max < iftype_num[iftype])
-					goto cont;
-				limits[j].max -= iftype_num[iftype];
-			}
-		}
+                if (!*beacon_int_gcd) {
+                        *beacon_int_gcd = wdev->beacon_interval;
+                        continue;
+                }
 
-		if (radar_detect != (c->radar_detect_widths & radar_detect))
-			goto cont;
+                if (wdev->beacon_interval == *beacon_int_gcd)
+                        continue;
 
-		if (radar_detect && c->radar_detect_regions &&
-		    !(c->radar_detect_regions & BIT(region)))
-			goto cont;
+                *beacon_int_different = true;
+                *beacon_int_gcd = gcd(*beacon_int_gcd, wdev->beacon_interval);
+        }
 
-		/* Finally check that all iftypes that we're currently
-		 * using are actually part of this combination. If they
-		 * aren't then we can't use this combination and have
-		 * to continue to the next.
-		 */
-		if ((all_iftypes & used_iftypes) != used_iftypes)
-			goto cont;
+        if (new_beacon_int && *beacon_int_gcd != new_beacon_int) {
+                if (*beacon_int_gcd)
+                        *beacon_int_different = true;
+                *beacon_int_gcd = gcd(*beacon_int_gcd, new_beacon_int);
+        }
+}
 
-		/* This combination covered all interface types and
-		 * supported the requested numbers, so we're good.
-		 */
+int cfg80211_validate_beacon_int(struct cfg80211_registered_device *rdev,
+                                 enum nl80211_iftype iftype, u32 beacon_int)
+{
+        /*
+         * This is just a basic pre-condition check; if interface combinations
+         * are possible the driver must already be checking those with a call
+         * to cfg80211_check_combinations(), in which case we'll validate more
+         * through the cfg80211_calculate_bi_data() call and code in
+         * cfg80211_iter_combinations().
+         */
+
+        if (beacon_int < 10 || beacon_int > 10000)
+                return -EINVAL;
+
+        return 0;
+}
 
-		(*iter)(c, data);
+int cfg80211_iter_combinations(struct wiphy *wiphy,
+                               struct iface_combination_params *params,
+                               void (*iter)(const struct ieee80211_iface_combination *c,
+                                            void *data),
+                               void *data)
+{
+        const struct ieee80211_regdomain *regdom;
+        enum nl80211_dfs_regions region = 0;
+        int i, j, iftype;
+        int num_interfaces = 0;
+        u32 used_iftypes = 0;
+    u32 beacon_int_gcd;
+        bool beacon_int_different;
+
+        /*
+         * This is a bit strange, since the iteration used to rely only on
+         * the data given by the driver, but here it now relies on context,
+         * in form of the currently operating interfaces.
+         * This is OK for all current users, and saves us from having to
+         * push the GCD calculations into all the drivers.
+         * In the future, this should probably rely more on data that's in
+         * cfg80211 already - the only thing not would appear to be any new
+         * interfaces (while being brought up) and channel/radar data.
+         */
+        cfg80211_calculate_bi_data(wiphy, params->new_beacon_int,
+                                   &beacon_int_gcd, &beacon_int_different);
+
+        if (params->radar_detect) {
+                rcu_read_lock();
+                regdom = rcu_dereference(cfg80211_regdomain);
+                if (regdom)
+                        region = regdom->dfs_region;
+                rcu_read_unlock();
+        }
+
+        for (iftype = 0; iftype < NUM_NL80211_IFTYPES; iftype++) {
+                num_interfaces += params->iftype_num[iftype];
+                if (params->iftype_num[iftype] > 0 &&
+                    !(wiphy->software_iftypes & BIT(iftype)))
+                        used_iftypes |= BIT(iftype);
+        }
+
+        for (i = 0; i < wiphy->n_iface_combinations; i++) {
+                const struct ieee80211_iface_combination *c;
+                struct ieee80211_iface_limit *limits;
+                u32 all_iftypes = 0;
+
+                c = &wiphy->iface_combinations[i];
+
+                if (num_interfaces > c->max_interfaces)
+                        continue;
+                if (params->num_different_channels > c->num_different_channels)
+                        continue;
+
+                limits = kmemdup(c->limits, sizeof(limits[0]) * c->n_limits,
+                                 GFP_KERNEL);
+                if (!limits)
+                        return -ENOMEM;
+
+                for (iftype = 0; iftype < NUM_NL80211_IFTYPES; iftype++) {
+                        if (wiphy->software_iftypes & BIT(iftype))
+                                continue;
+                        for (j = 0; j < c->n_limits; j++) {
+                                all_iftypes |= limits[j].types;
+                                if (!(limits[j].types & BIT(iftype)))
+                                        continue;
+                                if (limits[j].max < params->iftype_num[iftype])
+                                        goto cont;
+                                limits[j].max -= params->iftype_num[iftype];
+                        }
+                }
+
+                if (params->radar_detect !=
+                        (c->radar_detect_widths & params->radar_detect))
+                        goto cont;
+
+                if (params->radar_detect && c->radar_detect_regions &&
+                    !(c->radar_detect_regions & BIT(region)))
+                        goto cont;
+
+                /* Finally check that all iftypes that we're currently
+                 * using are actually part of this combination. If they
+                 * aren't then we can't use this combination and have
+                 * to continue to the next.
+                 */
+                if ((all_iftypes & used_iftypes) != used_iftypes)
+                        goto cont;
+
+                if (beacon_int_gcd) {
+                        if (c->beacon_int_min_gcd &&
+                            beacon_int_gcd < c->beacon_int_min_gcd)
+                                goto cont;
+                        if (!c->beacon_int_min_gcd && beacon_int_different)
+                                goto cont;
+                }
+
+                /* This combination covered all interface types and
+                 * supported the requested numbers, so we're good.
+                 */
+                (*iter)(c, data);
  cont:
-		kfree(limits);
-	}
+                kfree(limits);
+        }
 
-	return 0;
+        return 0;
 }
 EXPORT_SYMBOL(cfg80211_iter_combinations);
 
@@ -1601,21 +1651,18 @@ cfg80211_iter_sum_ifcombs(const struct ieee80211_iface_combination *c,
 }
 
 int cfg80211_check_combinations(struct wiphy *wiphy,
-				const int num_different_channels,
-				const u8 radar_detect,
-				const int iftype_num[NUM_NL80211_IFTYPES])
+                                struct iface_combination_params *params)
 {
-	int err, num = 0;
-
-	err = cfg80211_iter_combinations(wiphy, num_different_channels,
-					 radar_detect, iftype_num,
-					 cfg80211_iter_sum_ifcombs, &num);
-	if (err)
-		return err;
-	if (num == 0)
-		return -EBUSY;
+        int err, num = 0;
 
-	return 0;
+        err = cfg80211_iter_combinations(wiphy, params,
+                                         cfg80211_iter_sum_ifcombs, &num);
+        if (err)
+                return err;
+        if (num == 0)
+                return -EBUSY;
+
+        return 0;
 }
 EXPORT_SYMBOL(cfg80211_check_combinations);
 
@@ -1627,15 +1674,15 @@ int cfg80211_can_use_iftype_chan(struct cfg80211_registered_device *rdev,
 				 u8 radar_detect)
 {
 	struct wireless_dev *wdev_iter;
-	int num[NUM_NL80211_IFTYPES];
 	struct ieee80211_channel
 			*used_channels[CFG80211_MAX_NUM_DIFFERENT_CHANNELS];
 	struct ieee80211_channel *ch;
 	enum cfg80211_chan_mode chmode;
-	int num_different_channels = 0;
 	int total = 1;
 	int i;
-
+	struct iface_combination_params params = {
+                .radar_detect = radar_detect,
+	};
 	ASSERT_RTNL();
 
 	if (WARN_ON(hweight32(radar_detect) > 1))
@@ -1651,10 +1698,9 @@ int cfg80211_can_use_iftype_chan(struct cfg80211_registered_device *rdev,
 		return 0;
 	}
 
-	memset(num, 0, sizeof(num));
 	memset(used_channels, 0, sizeof(used_channels));
 
-	num[iftype] = 1;
+	params.iftype_num[iftype] = 1;
 
 	/* TODO: We'll probably not need this anymore, since this
 	 * should only be called with CHAN_MODE_UNDEFINED. There are
@@ -1667,14 +1713,14 @@ int cfg80211_can_use_iftype_chan(struct cfg80211_registered_device *rdev,
 	case CHAN_MODE_SHARED:
 		WARN_ON(!chan);
 		used_channels[0] = chan;
-		num_different_channels++;
+		params.num_different_channels++;
 		break;
 	case CHAN_MODE_EXCLUSIVE:
-		num_different_channels++;
+		params.num_different_channels++;
 		break;
 	}
 
-	list_for_each_entry(wdev_iter, &rdev->wdev_list, list) {
+	list_for_each_entry(wdev_iter, &rdev->wiphy.wdev_list, list) {
 		if (wdev_iter == wdev)
 			continue;
 		if (wdev_iter->iftype == NL80211_IFTYPE_P2P_DEVICE) {
@@ -1698,7 +1744,7 @@ int cfg80211_can_use_iftype_chan(struct cfg80211_registered_device *rdev,
 		 */
 		mutex_lock_nested(&wdev_iter->mtx, 1);
 		__acquire(wdev_iter->mtx);
-		cfg80211_get_chan_state(wdev_iter, &ch, &chmode, &radar_detect);
+		cfg80211_get_chan_state(wdev_iter, &ch, &chmode, &params.radar_detect);
 		wdev_unlock(wdev_iter);
 
 		switch (chmode) {
@@ -1714,23 +1760,22 @@ int cfg80211_can_use_iftype_chan(struct cfg80211_registered_device *rdev,
 
 			if (used_channels[i] == NULL) {
 				used_channels[i] = ch;
-				num_different_channels++;
+				params.num_different_channels++;
 			}
 			break;
 		case CHAN_MODE_EXCLUSIVE:
-			num_different_channels++;
+			params.num_different_channels++;
 			break;
 		}
 
-		num[wdev_iter->iftype]++;
+		params.iftype_num[wdev_iter->iftype]++;
 		total++;
 	}
 
-	if (total == 1 && !radar_detect)
+	if (total == 1 && !params.radar_detect)
 		return 0;
 
-	return cfg80211_check_combinations(&rdev->wiphy, num_different_channels,
-					   radar_detect, num);
+	return cfg80211_check_combinations(&rdev->wiphy, &params);
 }
 
 int ieee80211_get_ratemask(struct ieee80211_supported_band *sband,
@@ -1812,3 +1857,54 @@ EXPORT_SYMBOL(rfc1042_header);
 const unsigned char bridge_tunnel_header[] __aligned(2) =
 	{ 0xaa, 0xaa, 0x03, 0x00, 0x00, 0xf8 };
 EXPORT_SYMBOL(bridge_tunnel_header);
+
+bool cfg80211_is_gratuitous_arp_unsolicited_na(struct sk_buff *skb)
+{
+        const struct ethhdr *eth = (void *)skb->data;
+        const struct {
+                struct arphdr hdr;
+                u8 ar_sha[ETH_ALEN];
+                u8 ar_sip[4];
+                u8 ar_tha[ETH_ALEN];
+                u8 ar_tip[4];
+        } __packed *arp;
+        const struct ipv6hdr *ipv6;
+        const struct icmp6hdr *icmpv6;
+
+        switch (eth->h_proto) {
+        case cpu_to_be16(ETH_P_ARP):
+                /* can't say - but will probably be dropped later anyway */
+                if (!pskb_may_pull(skb, sizeof(*eth) + sizeof(*arp)))
+                        return false;
+
+                arp = (void *)(eth + 1);
+
+                if ((arp->hdr.ar_op == cpu_to_be16(ARPOP_REPLY) ||
+                     arp->hdr.ar_op == cpu_to_be16(ARPOP_REQUEST)) &&
+                    !memcmp(arp->ar_sip, arp->ar_tip, sizeof(arp->ar_sip)))
+                        return true;
+                break;
+        case cpu_to_be16(ETH_P_IPV6):
+                /* can't say - but will probably be dropped later anyway */
+                if (!pskb_may_pull(skb, sizeof(*eth) + sizeof(*ipv6) +
+                                        sizeof(*icmpv6)))
+                        return false;
+
+                ipv6 = (void *)(eth + 1);
+                icmpv6 = (void *)(ipv6 + 1);
+
+                if (icmpv6->icmp6_type == NDISC_NEIGHBOUR_ADVERTISEMENT &&
+                    !memcmp(&ipv6->saddr, &ipv6->daddr, sizeof(ipv6->saddr)))
+                        return true;
+                break;
+        default:
+                /*
+                 * no need to support other protocols, proxy service isn't
+                 * specified for any others
+                 */
+                break;
+        }
+
+        return false;
+}
+EXPORT_SYMBOL(cfg80211_is_gratuitous_arp_unsolicited_na);
-- 
1.9.1

